File: Python/ByRound/2092/2092_F_Andryusha_and_CCB.py
# Problem: CF 2092 F - Andryusha and CCB
# https://codeforces.com/contest/2092/problem/F

"""
F. Andryusha and CCB
Time Complexity: O(n^2) in worst case, but with optimizations it's closer to O(n * sqrt(n))
Space Complexity: O(n)

This solution computes for each prefix of the binary string, how many values of k 
are possible such that we can split the prefix into exactly k parts with equal beauty.

The beauty of a string is defined as the number of adjacent differing characters.
For a valid split:
1. The total beauty must be divisible by k
2. Each part will have the same beauty value
3. We need to partition the string such that each part's beauty is maintained

We use segment tree concepts and prefix sum techniques to efficiently compute 
these values across all prefixes.
"""

import sys


def solve():
    data = sys.stdin.read().split()
    t = int(data[0])
    idx = 1
    out_lines = []
    for _ in range(t):
        n = int(data[idx])
        idx += 1
        s = data[idx]
        idx += 1
        a = []  # Store lengths of consecutive segments of same character
        curr = 1
        for i in range(1, n):  # Traverse the string to group consecutive same characters
            if s[i] == s[i - 1]:
                curr += 1
            else:
                a.append(curr)
                curr = 1
        a.append(curr)
        sz = len(a)
        left = [0] * sz
        right = [0] * sz
        if sz > 0:
            # left[i] and right[i] define range of indices in the final array that corresponds to segment i
            left[0] = 0
            right[0] = a[0] - 1
            for i in range(1, sz):
                left[i] = right[i - 1] + 1
                right[i] = left[i] + a[i] - 1

        ans = [0] * n
        ans_local = ans
        left_local = left
        right_local = right
        
        # First pass: assign initial values based on linear increase per segment
        for i in range(sz):
            li = left_local[i]
            ri = right_local[i]
            const = 1 - i
            for j in range(li, ri + 1):
                ans_local[j] += j + const

        # Second phase: handle dependencies using difference array technique
        if sz > 1:
            a_local = a
            add = [0] * sz
            add_local = add
            for m in range(1, sz):  # Loop through possible gap sizes (m)
                l = m
                r = m
                while l < sz:
                    add_local[l] += 1
                    nr = r + 1
                    if nr < sz:
                        add_local[nr] -= 1
                    if a_local[l] == 1:  # If segment length is 1, move by m+1 steps
                        l += m + 1
                    else:  # Otherwise move by m steps
                        l += m
                    r += m + 1
            
            # Compute prefix sum to apply the updates
            pref = 0
            for i in range(sz):
                pref += add_local[i]
                if pref != 0:
                    li = left_local[i]
                    ri = right_local[i]
                    val = pref
                    for j in range(li, ri + 1):
                        ans_local[j] += val

        out_lines.append(" ".join(str(ans_local[i]) for i in range(n)))
    sys.stdout.write("\n".join(out_lines))


if __name__ == "__main__":
    solve()


# https://github.com/VaHiX/codeForces/

File: CPP/ByTier/F/2092_F_Andryusha_and_CCB.cpp
// Problem: CF 2092 F - Andryusha and CCB
// https://codeforces.com/contest/2092/problem/F

/*
F. Andryusha and CCB
Algorithm: Segment tree / prefix analysis / group compression technique
Time Complexity: O(n) amortized per test case due to repeated use of precomputed data and efficient looping
Space Complexity: O(n)

This code processes a binary string and for each prefix, calculates how many ways it can be split into k equal beauty substrings.
Uses grouping of consecutive same characters and difference array techniques to efficiently compute counts.

Key parts:
- Grouping consecutive characters into segments with cd[] indicating if segment is repeated (equal)
- Using difference array qzc[] for efficient interval updates
- Final computation of valid k values using a prefix sum style approach

*/

#include <iostream>

#pragma GCC optimize(3, "Ofast", "inline")
using namespace std;
#define ll long long
const int MAXN = 1e6 + 2;
const int MOD = 998244353;
bool cd[MAXN];      // Tracks if a group is repeated (consecutive same characters)
int qzc[MAXN];      // Difference array for interval updates
char s[MAXN];
int main() {
  ios::sync_with_stdio(0);
  cin.tie(0);
  cout.tie(0);
  int casenum;
  cin >> casenum;
  while (casenum--) {
    int n;
    cin >> n;
    cin >> s;
    int cnt = 0;     // Number of groups
    cd[0] = 0;       // First group
    for (int i = 1; i < n; i++) {
      if (s[i] == s[i - 1])
        cd[cnt] = 1;   // True if same as previous
      else
        cnt++, cd[cnt] = 0;  // Move to next group
    }
    for (int i = 0; i <= cnt; i++)
      qzc[i] = 0;    // Reset difference array
    
    // Update qzc with segment intervals using sliding window
    for (int i = 1; i < n; i++) {
      int l = 0, r = 0;
      while (l <= cnt) {
        if (l + i > cnt)
          break;
        qzc[l + i]++;           // Increment start of interval
        if (r + i + 1 <= cnt)   // Decrement end+1 of interval
          qzc[r + i + 1]--;
        r += i + 1;             // Next interval start
        if (cd[l + i])          // If new group is repeated, advance by i only
          l += i;
        else                    // Otherwise advance by i + 1
          l += i + 1;
      }
    }
    
    int cnt0 = 0;     // Current index in groups
    int now = 0;
    cout << 1 << " ";    // k=1 always valid
    
    for (int i = 1; i < n; i++) {
      if (s[i] != s[i - 1]) // If new group begins, update current count
        now += qzc[++cnt0];
      cout << now + i - cnt0 + 1 << ' ';  // Final count for this prefix
    }
    cout << endl;
  }
}


// https://github.com/VaHiX/codeForces/